### PR TITLE
Introduced ignore attribute to JsonKey.

### PR DESCRIPTION
**Pull request by AlexanderJohr**
Hello Kevin Moore, from all the serialization libraries for dart out there, yours was always my favorite. But for me, it was lacking 2 major features: an ignore annotation and keep initial values if the deserialized field is null. I created two branches in this fork which include these features. With them, it's all I wanted from a serialization library for dart for a long time and I use it with joy already, but I want to share my joy with the community.

I want to add examples, tests, how to guides in the readme and so on. I also have to run the tests, which is difficult for me under windows, because the tools don't work so well and I'm lacking time to do it on Linux.

The reason for those pull requests is that I want to get in touch with you first and get feedback on how to change or rename certain things.

This pull request is for the ignore attribute for JsonKey. When set to true the generator excludes the field.

```dart
@JsonSerializable(includeIfNull: false, nullable: false)
class Column extends _$ColumnSerializerMixin {
// [...]
// ColumnSettings is serializable
ColumnSettings settings = new ColumnSettings();

// ColumnSettings is already serializable and includes witdh.
// Without 'ignore' the generator would generate the same field for ColumnSettings and Column 
@JsonKey(ignore: true)
num get width => settings.width;
set width (num value) =>settings.width = value;
// [...]
}
```

What is also does is:
- exclude private fields
- refactored _writeFactory to do only one thing. Extracted the part where it fetches the final fields.

What I want to know is, if you recommend:

- to use another annotation to ignore a field instead of an ignore attribute in JsonKey, because ig ignored why call it JsonKey?
- to reenable private fields
- to throw an error if someone tries to use a private field in a constructor if private fields should be excluded? Your awesome library even handles deserializing into named and positioned constructor parameters (I don't know any library that doesn't complain if there isn't a parameterless default constructor.). But if the parameter assigns to a private field it must be possible to deserialize that field too.

### CODE REVIEW

File: json_serializable/lib/src/json_serializable_generator.dart [hunk 1]
```diff
@@ -315,10 +320,11 @@ void $toJsonMapHelperName(String key, dynamic value) {
   }
 
   /// Returns the set of fields that are not written to via constructors.
-  Set<FieldElement> _writeFactory(
+  _writeFactory(
```
Comment (kevmoo) [inline json_serializable/lib/src/json_serializable_generator.dart]: change this to `void`

**Comment by kevmoo** [comment]
Please add tests for this as well – even just changing one of the examples in in `test/` to use this key to show how it works...

**Comment by AlexanderJohr** [comment]
Ok, I will run the tests soon on a non-windows machine soon and fix them and also add new ones for the ignore feature.

But apart from that, do you have feedback for me regarding my questions at the end of my pull request comment?

**Comment by kevmoo** [comment]
Looking better – check out this guy - https://travis-ci.org/dart-lang/json_serializable/jobs/352621359#L561

**Comment by kevmoo** [comment]
@AlexanderJohr ?

**Comment by AlexanderJohr** [comment]
@kevmoo Yes, looking into it this weekend!

**Comment by AlexanderJohr** [comment]
I fixed the bugs and made the tests run green.
I also added a test showing the usage of the ignore field and two more tests to ensure final fields are included when set in constructor and excluded when not.

**Comment by AlexanderJohr** [comment]
Sorry, I oversaw that dartfmt needs to be run on that one file.

**Comment by AlexanderJohr** [comment]
I had to rewrite the git history because I committed on different systems with different email addresses without noticing. I thought this would remove the CLA waring but it seems that it made it worse. Should I just fork again, apply the changes and create a new Pull Request with the same Github user?

**Comment by kevmoo** [comment]
> Should I just fork again, apply the changes and create a new Pull Request with the same Github user?

That's fine!

**Comment by AlexanderJohr** [comment]
Created new Pull Request with CLA signed commit authors: #118

**Comment by AlexanderJohr** [comment]
Created new Pull Request with CLA signed commit authors: #118

**Pull request closed without merging**